src/image.rs

```diff
@@ -133,6 +133,23 @@ impl ImageFormat {
         inner(path.as_ref())
     }
 
+    /// Return if the current path is can be read by the lib.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use image::ImageFormat;
+    ///
+    /// let result = ImageFormat::can_read("images/ferris.png");
+    /// assert_eq!(result, true);
+    /// ```
+    #[inline]
+    pub fn can_read<P>(path: P) -> bool where P : AsRef<Path> {
```

Comment:
I thought we were going to make this have signature `pub fn can_read(&self) -> bool`?

Reply:
oh I misunderstood it then, I will update it soon 👍

Reply:
While trying to implement it as `pub fn can_read(&self) -> bool` I'm even more confused on what to implement exactly, sorry for this but I will need a more concrete example on what to do.

From what I read from your first post, the problem we're trying to solve is filtering the results of a glob search into just the paths that are supported by the lib in a way that needs not to be modified if a new type is supported, which was way I implemented it as `pub fn can_read<P>(path: P) -> bool where P : AsRef<Path>` which is still wrong as it doesn't check for the existance of a decoder for the corresponding ImageFromat, only that the ImageFromat exists in the crate.

Now for it to be `pub fn can_read(&self) -> bool` it implies that you already have the ImageFromat (which is odd given you're trying to check from paths, this means you already had to run something to map the path to the ImageFromat) and are checking for support for encode/decode right? So what I would have to implement would be a way to match self to a decoder/encoder for can_read/can_write respectively.

However since the way it is done in `pub fn load<R: BufRead + Seek>(r: R, format: ImageFormat) -> ImageResult<DynamicImage>` it's trough a match statement I see no other option other than copy paste the match & update it to return what I can see maps to a decoder now, the problem with this is that if a current format is added without a decoder & then the decoder is added latter this match can get out of sync with the libs functionality, which would be troublesome to maintain.

Is this really what we want for this, or is it something else?

Thank you for the patience & all the help!

Reply:
> Now for it to be pub fn can_read(&self) -> bool it implies that you already have the ImageFromat (which is odd given you're trying to check from paths, this means you already had to run something to map the path to the ImageFromat) and are checking for support for encode/decode right?

Yes. To go from a path / file extension to an ImageFormat we have [`from_path`](https://docs.rs/image/0.23.12/image/enum.ImageFormat.html#method.from_path) / [`from_extension`](https://docs.rs/image/0.23.12/image/enum.ImageFormat.html#method.from_extension).

> I see no other option other than copy paste the match & update it to return what I can see maps to a decoder now, the problem with this is that if a current format is added without a decoder & then the decoder is added latter this match can get out of sync with the libs functionality, which would be troublesome to maintain.

Yes, I was just imagining a match statement. You are right that this adds one more place to get out of sync, but if we were able to do the kind of introspection necessary to avoid that, we probably wouldn't need this function in the first place :wink: